### PR TITLE
Add MSLIC (Manifold SLIC)

### PR DESCRIPTION
**Pull request by cbalint13**
This PR adds **MSLIC** [1] superpixels for image segmentation.

Some infos:
- Implementation here is properly cited in respect with author.
- Implementation is derived using BSD, author [2] agreed for the license terms used here.
- Implementation booth follows opencv's coding standard but author original code.
- Implementation behave as original one but speed is improved using opencv parallelism.
- Implementation can handle CV_{XX}C{Y} image types ( {Y} = multispectral ).
- Implementation have documented flexible parameters and follows original demonstrator.

[1] [CVPR2016 - MSLIC](http://www.cv-foundation.org/openaccess/content_cvpr_2016/html/Liu_Manifold_SLIC_A_CVPR_2016_paper.html)
[2]  [Yong-Jin Liu](http://cg.cs.tsinghua.edu.cn/people/~Yongjin/Yongjin.htm) `<liuyongjin [at] tsinghua [dot] edu [dot] cn>`

Author is invited to this PR review, he may interact within process.

slic-demo

### CODE REVIEW

File: modules/ximgproc/include/opencv2/ximgproc/slic.hpp [hunk 1]
```diff
@@ -138,21 +140,22 @@ class CV_EXPORTS_W SuperpixelSLIC : public Algorithm
```
Comment (sovrasov) [inline modules/ximgproc/include/opencv2/ximgproc/slic.hpp]: Brief description of this function should be changed to something like `Initializes a SuperpixelSLIC object.` It's an old copy-paste error.

File: modules/ximgproc/include/opencv2/ximgproc/slic.hpp [hunk 2]
```diff
@@ -138,21 +140,22 @@ class CV_EXPORTS_W SuperpixelSLIC : public Algorithm
...
+computing iterations over the given image. For enanched results it is recommended for color images to
+preprocess image with little gaussian blur using a small 3 x 3 kernel and additional conversion into
+CieLAB color space. An example of SLIC versus SLICO and MSLIC is ilustrated in the following picture.
 
 ![image](pics/superpixels_slic.png)
 
  */
 
-    enum SLIC { SLIC = 100, SLICO = 101 };
+    enum SLIC { SLIC = 100, SLICO = 101, MSLIC = 102 };
```
Comment (sovrasov) [inline modules/ximgproc/include/opencv2/ximgproc/slic.hpp]: Please, move this enum to the line 63 in this file to let Doxygen process it correctly.
Now description of the `createSuperpixelSLIC` function is empty.

**Comment by sovrasov** [comment]
To enable switching to MSLIC in sample via trackbar 1 should be replaced with 2 [here](https://github.com/cbalint13/opencv_contrib/blob/33925144b4f2c776841ee08844cb2b9cf7f62753/modules/ximgproc/samples/slic.cpp#L66).

**Comment by cbalint13** [comment]
@sovrasov ,

* All done as requested.

**Comment by EdwardMa12593** [comment]
Would it be possible for you to provide MALTAB version of the code?

**Comment by cbalint13** [comment]
@EdwardMa12593 ,

  * Original code was C++ .
  * You can ask the very author (e-mail) regarding if he have matlab .

Personally can't help with matlab code.

**Pull request merged**